File: 0053_maximum_subarray/max_subarray.js
/**
 * Time: O(n)
 * Space: O(1)
 */
const maxSubArray = nums => {
  if (!nums.length) return 0
  for (let i = 1; i < nums.length; i++)
    if (nums[i - 1] > 0) nums[i] = nums[i] + nums[i - 1]
  return Math.max(...nums)
}

File: 0053_maximum_subarray/max_subarray.java
class Solution {
    // Time: O(n)
    // Space: O(1)
    public int maxSubArray(int[] nums) {
        if (nums.length == 0)
            return 0;
        int maxSoFar = nums[0], maxCurr = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int curr = nums[i];
            maxCurr = Math.max(curr, maxCurr + curr);
            maxSoFar = Math.max(maxCurr, maxSoFar);
        }
        return maxSoFar;
    }
}